apps/src/code-studio/components/AudioRecorder.jsx

```diff
@@ -61,21 +62,34 @@ export default class AudioRecorder extends React.Component {
   };
 
   saveAudio = (blob) => {
-    assetsApi.putAsset(this.state.audioName + ".mp3", blob,
-    (xhr) => {
-      this.setState({audioName: ""});
-      this.props.onUploadDone(JSON.parse(xhr.response));
-      this.props.afterAudioSaved(AudioErrorType.NONE);
-    }, error => {
-      console.error(`Audio Failed to Save: ${error}`);
-      this.props.afterAudioSaved(AudioErrorType.SAVE);
-    });
+    if (!this.state.cancelling) {
+      assetsApi.putAsset(this.state.audioName + ".mp3", blob,
+      (xhr) => {
+        this.setState({audioName: ""});
+        this.props.onUploadDone(JSON.parse(xhr.response));
+        this.props.afterAudioSaved(AudioErrorType.NONE);
+      }, error => {
+        console.error(`Audio Failed to Save: ${error}`);
+        this.props.afterAudioSaved(AudioErrorType.SAVE);
+      });
+    }
   };
 
   onNameChange = (event) => {
     this.setState({audioName: event.target.value});
   };
 
+  onCancel = () => {
+    this.setState({audioName: "", recording: false, cancelling: true}, () => {
+      this.props.afterAudioSaved(AudioErrorType.NONE);
+      // Only stop recording if it's been started
+      if (this.recorder.state !== "inactive") {
+        this.recorder.stop();
+      }
+      this.setState({cancelling: false});
+    });
```

Comment:
There are a couple of things we might be able to clean up here.  One of them is that the `canceling` state isn't ever used in rendering, so it could just be an instance variable on the component, which in turn eliminates some of the asynchronous nature of this operation.
```js
onCancel = () => {
  this.setState({audioName: "", recording: false});
  this.canceling = true;
  this.props.afterAudioSaved(AudioErrorType.NONE);
  if (this.recorder.state !== "inactive") {
    this.recorder.stop();
  }
  this.canceling = false;
};
```
It might make sense to move the one `setState` call to the end of the method; in practice it doesn't happen until after the method anyway, at that point.

The second thing here is both a code smell and maybe risk of a bug - they way we're modifying the behavior of `this.recorder.stop()` by setting some instance state before and after it.

I say it's a code smell because it's stateful where it feels like we should get away with not using state - I'm not sure what this looks like yet, but it'd be nicer to be able to say something like
```js
this.stopRecording({save: false});
```
The possible bug is that we are trusting here that `this.recorder.stop()` is a synchronous operation, but the relevant behavior is implemented as an `onstop` callback and so it's not obviously guaranteed that it is synchronous.  Maybe in practice it is; that's an implementation detail worth calling out with a comment if so.